### PR TITLE
Add serial number generation feature

### LINKED ISSUE #73: Implement serial number generator

### PR DESCRIPTION
**Pull request by kaichaoang**
This feature will ease future method or command calls by mapping a stock to its unique serial number.

### CODE REVIEW

File: src/main/java/seedu/stock/logic/parser/AddCommandParser.java [hunk 1]
```diff
@@ -37,7 +37,7 @@ public AddCommand parse(String args) throws ParseException {
         }
 
         Name name = ParserUtil.parseName(argMultimap.getValue(PREFIX_NAME).get());
-        SerialNumber serialNumber = SerialNumber.generateSerialNumber();
+        SerialNumber serialNumber = SerialNumber.generateDefaultSerialNumber();
```
Comment (amylzting) [inline src/main/java/seedu/stock/logic/parser/AddCommandParser.java:40]: Would it be better if there was no need to generate default serial number here? As I feel it is an extra step before we get the actual serial number for the stock. I may have an idea of how to do it, and can help to work on it!

File: src/main/java/seedu/stock/model/SerialNumberSetsBook.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    /**
+     * Resets the existing data of this {@code SerialNumberSetsBook} with {@code newData}.
+     */
+    public void resetData(ReadOnlySerialNumberSetsBook newData) {
+        requireNonNull(newData);
+
+        setSerialNumberSets(newData.getSerialNumberSetsList());
+    }
+
+    //// person-level operations
```
Comment (amylzting) [inline src/main/java/seedu/stock/model/SerialNumberSetsBook.java:62]: Perhaps could remove this line as it seems it is not relevant!

File: src/main/java/seedu/stock/model/UserPrefs.java [hunk 3]
```diff
@@ -81,6 +93,7 @@ public String toString() {
         StringBuilder sb = new StringBuilder();
         sb.append("Gui Settings : " + guiSettings);
         sb.append("\nLocal data file location : " + stockBookFilePath);
+        sb.append("\nLocal serial number data file location : " + serialNumberSetsBookFilePath);
```
Comment (amylzting) [inline src/main/java/seedu/stock/model/UserPrefs.java:96]: Perhaps it is not needed to show the path to the serial number sets book?

File: src/main/java/seedu/stock/model/stock/AccQuantity.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    @Override
+    public String toString() {
+        return accQuantity;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof AccQuantity // instanceof handles nulls
+                && accQuantity.equals(((AccQuantity) other).accQuantity)); // state check
```
Comment (amylzting) [inline src/main/java/seedu/stock/model/stock/AccQuantity.java]: This line is pretty long. Perhaps could break it down to improve code readability?

File: src/main/java/seedu/stock/model/stock/SerialNumberSet.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+     * Returns true if both serialNumberSet is the same.
+     */
+    public boolean isSameSerialNumberSet(SerialNumberSet otherSerialNumberSet) {
+        if (otherSerialNumberSet == this) {
+            return true;
+        }
+
+        return otherSerialNumberSet != null
+                && otherSerialNumberSet.getSource().equals(getSource())
+                && otherSerialNumberSet.getAccQuantity().equals(getAccQuantity());
```
Comment (amylzting) [inline src/main/java/seedu/stock/model/stock/SerialNumberSet.java]: Perhaps could break down into more readable statements. Seen a few times!

File: src/main/java/seedu/stock/model/stock/SerialNumber.java [hunk 6]
```diff
@@ -42,8 +45,8 @@ public String getSerialNumber() {
      *
      * @return SerialNumber of the object.
      */
-    public static SerialNumber generateSerialNumber() {
-        return new SerialNumber("12");
+    public static SerialNumber generateDefaultSerialNumber() {
```
Comment (amylzting) [inline src/main/java/seedu/stock/model/stock/SerialNumber.java:47]: As mentioned, perhaps it is possible to remove this method!

File: src/main/java/seedu/stock/model/stock/UniqueSerialNumberSetList.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    /**
+     * Adds a serial number set to the list.
+     * The serial number set's source must not already exist in the list.
+     */
+    public void add(SerialNumberSet toAdd) {
+        requireNonNull(toAdd);
+        if (contains(toAdd)) {
+            throw new DuplicateStockException();
```
Comment (amylzting) [inline src/main/java/seedu/stock/model/stock/UniqueSerialNumberSetList.java:47]: Perhaps could make a new exception class for the duplicate source?

File: src/main/java/seedu/stock/model/stock/exceptions/DuplicateSerialNumberSetException.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package seedu.stock.model.stock.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate Serial Number set (Serial number sets
+ *     are considered duplicates if they have the same identity).
+ */
+public class DuplicateSerialNumberSetException extends RuntimeException {
+    public DuplicateSerialNumberSetException() {
+        super("Operation would result in duplicate stock.");
```
Comment (amylzting) [inline src/main/java/seedu/stock/model/stock/exceptions/DuplicateSerialNumberSetException.java]: Perhaps "stock" should be "serial number set"?

**Review by amylzting — COMMENTED** [review]
Overall, LGTM 👍

File: src/main/java/seedu/stock/logic/parser/AddCommandParser.java (hunk 1 above)
Comment (Ashley-Lau) [reply src/main/java/seedu/stock/logic/parser/AddCommandParser.java:40]: Where do you suggest then? The other location I can think of is stock class.

File: src/main/java/seedu/stock/model/SerialNumberSetsBook.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * Returns true if a SerialNumberSet with the same identity as {@code serialNumberSet} exists
+     *     in the SerialNumberSetsBook.
```
Comment (Ashley-Lau) [inline src/main/java/seedu/stock/model/SerialNumberSetsBook.java:66]: Is there an extra tab here?

File: src/main/java/seedu/stock/model/stock/AccQuantity.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
+package seedu.stock.model.stock;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.stock.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a source's accumulated quantity.
+ */
+public class AccQuantity {
```
Comment (Ashley-Lau) [inline src/main/java/seedu/stock/model/stock/AccQuantity.java]: I think you shouldn't use the short form of accumulated here for the class name?

File: src/main/java/seedu/stock/model/stock/AccQuantity.java (hunk 4 above)
Comment (Ashley-Lau) [reply src/main/java/seedu/stock/model/stock/AccQuantity.java]: I think this is fine.

**Review by Ashley-Lau — COMMENTED** [review]
Some changes you might want to make before merging!

File: src/main/java/seedu/stock/model/stock/AccQuantity.java (hunk 10 above)
Comment (AudreyFelicio) [reply src/main/java/seedu/stock/model/stock/AccQuantity.java]: I agree with Ashley here.

**Review by AudreyFelicio — APPROVED** [review]
The code looks neat. The serial number generator works after being tested. The default serial number and code quality issues aside, I think this can be merged already. Improvements can be made after merging this.

File: src/main/java/seedu/stock/model/stock/exceptions/DuplicateSerialNumberSetException.java (hunk 8 above)
Comment (kaichaoang) [reply src/main/java/seedu/stock/model/stock/exceptions/DuplicateSerialNumberSetException.java]: Noted!

File: src/main/java/seedu/stock/model/SerialNumberSetsBook.java (hunk 9 above)
Comment (kaichaoang) [reply src/main/java/seedu/stock/model/SerialNumberSetsBook.java:66]: I think I read that for javadocs next line should be 4 spaces, let me confirm again.

File: src/main/java/seedu/stock/model/stock/AccQuantity.java (hunk 10 above)
Comment (kaichaoang) [reply src/main/java/seedu/stock/model/stock/AccQuantity.java]: Noted!

**Pull request merged**